lib/spack/spack/test/concretize.py

```diff
@@ -455,22 +459,72 @@ def test_concretize_two_virtuals_with_dual_provider_and_a_conflict(self):
         with pytest.raises(spack.error.SpackError):
             s.concretize()
 
+    @pytest.mark.only_clingo(
+        "Optional compiler propagation isn't deprecated for original concretizer"
+    )
+    @pytest.mark.parametrize("unify", [True, False, "when_possible"])
+    def test_concretize_environment_propagated_disabled_variant(
+        self, unify, tmpdir, mutable_mock_env_path
+    ):
+        """Ensure that variants are propagated in a concrete environment"""
+        path = tmpdir.join("spack.yaml")
+
+        with tmpdir.as_cwd():
+            with open(str(path), "w") as f:
+                f.write(
+                    """\
+spack:
+  specs:
+    - ascent ~~shared +adios2
+"""
+                )
+
+        SpackCommand("env")("create", "test", str(path))
+
+        test = spack.environment.read("test")
+        test.unify = unify
+        test.concretize()
+
+        for spec in test.specs_by_hash.values():
+            for dep in spec.dependencies():
+                if dep.name == "adios2":
+                    assert dep.satisfies("~shared")
+                    assert dep.satisfies("^bzip2 ~shared")
+
     @pytest.mark.only_clingo(
         "Optional compiler propagation isn't deprecated for original concretizer"
     )
     def test_concretize_propagate_disabled_variant(self):
         """Test a package variant value was passed from its parent."""
-        spec = Spec("hypre~~shared ^openblas")
+        spec = Spec("ascent~~shared +adios2")
         spec.concretize()
 
-        assert spec.satisfies("^openblas~shared")
+        for dep in spec.traverse():
+            assert dep.satisfies("~shared")
 
+    @pytest.mark.only_clingo(
+        "Optional compiler propagation isn't deprecated for original concretizer"
+    )
     def test_concretize_propagated_variant_is_not_passed_to_dependent(self):
         """Test a package variant value was passed from its parent."""
-        spec = Spec("hypre~~shared ^openblas+shared")
+        spec = Spec("ascent~~shared +adios2 ^adios2+shared")
+        spec.concretize()
+
+        assert spec.satisfies("^adios2+shared")
+        assert spec.satisfies("^bzip2~shared")
```

Comment:
This is the only assertion failing in `develop` if I keep the tests from this PR and revert the changes in core.